### DISCUSSION TITLE
Image | Radxa ROCK 3 Model A

### DISCUSSION THREAD

**Discussion by cbake886** [MichaIng/DietPi#5168] [Image Request] [answered]
I am wondering if we can get a port of DietPi for one of Radxa's new SBCs, the ROCK 3 Model A (https://wiki.radxa.com/Rock3/hardware/3a)?

**ROCK 3A is basically a little brother of ROCK Pi 4, with some improvements from the experience on ROCK Pi 4, the main features are:**

- RK3568 based, Quad A55 up to 2GHz, Mali G52 GPU, 0.8T NPU
- pluggable eMMC module, 2/4/8GB LPDDR4 3200MT/s, dram controller frequency up to 1560MHz
- QC/PD power support, with a new generation QC/PD protocol IC
- Two USB2 same as ROCK Pi 4, two USB3(1x OTG+1x HOST), the two USB3 can be configured in SATA mode with a SATA breakout cable
- GbE with PoE support
- M.2 M key for NVMe with PCIe 3.0 x2 lanes
- M.2 E key for connectivity with PCIe 2.0 x1/SDIO/UART, support WiFi 6 cards
- 1x MIPI DSI, 1x MIPI CSI, compatible with ROCK Pi 4
- 40P GPIO mostly compatible with ROCK Pi 4

**Comment by MichaIng** [comment]
Many thank for your request.

Indeed that would be great. I'll ask Radxa for a developer sample.

**Reply by cbake886** [reply]
Any updates on this?

**Reply by Joulinar** [reply]
As far as I know we never received a single sample from Radxa. Even if they promised more than once.

**Reply by MichaIng** [reply]
I didn't ask for ROCK3 samples because the Radxa Zero's didn't arrive. However, there is a supported Armbian image, so using that as basis is should be easy to implement without testing our end necessarily required: https://www.armbian.com/rock-3a/

**Reply by cbake886** [reply]
If you can put something together I can test. Let me know. Thanks. 

Edit: I donated a few cups of coffee to help move this along. :)

**Reply by cbake886** [reply]
I installed Armbian 22.08 Jammy Kernel 5.19.y from https://www.armbian.com/rock-3a/ and ran your "dietpi-installer" script. Had to choose "Generic device". Only error i got was:

`Err:4 https://deb.debian.org/debian bookworm InRelease │ │ The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 648ACFD622F3D138 │ │ NO_PUBKEY 0E98404D386FA1D9`

I opened a subshell and ran the following script to resolve:

````shell
#!/bin/sh -e
tmp="$(mktemp)"
sudo apt-get update 2>&1 | sed -En 's/.*NO_PUBKEY ([[:xdigit:]]+).*/\1/p' | sort -u > "${tmp}"
cat "${tmp}" | xargs sudo gpg --keyserver "hkps://keyserver.ubuntu.com:443" --recv-keys  # to /usr/share/keyrings/*
cat "${tmp}" | xargs -L 1 sh -c 'sudo gpg --yes --output "/etc/apt/trusted.gpg.d/$1.gpg" --export "$1"' sh  # to /etc/apt/trusted.gpg.d/*
rm "${tmp}"
````


Upon reboot i got the following:

```
The programs included with the Debian GNU/Linux system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.

Debian GNU/Linux comes with ABSOLUTELY NO WARRANTY, to the extent
permitted by applicable law.
-dash: 2: /etc/profile.d/01-locale-fix.sh: /usr/bin/locale-check: not found
```

Tried to rebuild locale but that did not get rid of error. Had just a # as a prompt. So i installed zsh and set hostname to DietPi per /etc/hosts.

Does not seem as if the DietPi install script worked.

**Reply by nonnorm** [reply]
You'll have to use the Bullseye image, not Jammy.

**Comment by Joulinar** [comment]
DietPi is based on Debian and not Ubuntu. Therefore, you would need to use the Bullseye image and not Jammy https://redirect.armbian.com/region/EU/rock-3a/Bullseye_edge

Ah and you need to use DietPi Installer from development branch as we have a small issue on master that will create root user with incorrect dash shell.

**Reply by MichaIng** [reply]
Right, actually the installer should have exited automatically on distro detection, when not finding a supported Debian version 🤔.

**Reply by cbake886** [reply]
It gave me to option to select bookworm.

**Reply by MichaIng** [reply]
It should not have exited before the target distro selection is offered here: https://github.com/MichaIng/DietPi/blob/master/.build/images/dietpi-installer#L176

Only Bookworm was offended? Then this Ubuntu image seems to contain an `/etc/debian_version` file containing `bookworm/sid` :thinking:.

**Reply by cbake886** [reply]
Yea it warned me about using bookworm that its bleeding edge.(when i was using the ubuntu image*)

I used the dev branch install script and the bullseye image that Joulinar linked above. Selected "generic device" when installing and it seems to be running fine. I have docker with portainer + ELK stack running with great performance. Here is a dump from dmesg attached.

[dmesg_dump.txt](https://github.com/MichaIng/DietPi/files/9711588/dmesg_dump.txt)

**Comment by MichaIng** [comment]
PR up: #5823
It has a FAT boot partition and GPT partition table like the Armbian image, to work around issues with SPI bootloader (as far as I understood). It uses "edge" kernel/u-boot branch, i.e. mainline kernel which will be upgraded not only to new LTS versions but e.g. Linux 6.0, 6.1 etc. Once next Linux LTS release is up, I guess Armbian will provide "current" branch kernel/u-boot packages, so it makes sense to switch then.

The first testing image will be also DietPi `dev` branch, so only to see whether boot and main functionality works, i.e. there is no rough bug in build code. Once this is verified, I'll release `master` branch images, also after the DietPi v8.10 release tonight.

**Reply by MichaIng** [reply]
Images ready: https://dietpi.com/downloads/images/

**Comment by MichaIng** [comment, marked as answer]
There was a bug in our last image, not booting because of the rootfs UUID wasn't updated in our build scripts. Solved with latest image, however, would still be great if someone could verify that all works as expected: https://dietpi.com/#download

**Reply by Fgabz** [reply]
I've tried the last image and unfortunately it didn't boot. I didn't have time right now to get some logs. I will try later

**Reply by MichaIng** [reply]
Thanks for testing. Found and fixed it: https://github.com/MichaIng/DietPi/commit/3c9ba5b
Fixed images uploaded.

**Reply by Fgabz** [reply]
Thanks looks like it's working. I was able to boot and log in SSH

**Reply by MichaIng** [reply]
Great, again, thanks for testing!

**Comment by MichaIng** [comment]
@cbake886 @Fgabz
Has one of you time and mood to test a new image? I want to check whether it can boot without a dedicated boot partition, or even with MBR partition table instead of GPT. Image with MBR + single ext4 partition would be here: https://dietpi.com/downloads/images/testing/
If this does not boot, I'd create another one with GPT partition table but still a single ext4 filesystem only. If you have a UART adapter, that output would of course be very helpful.

**Reply by RadxaYuntian** [reply]
Replying here so #6220 won't be hijacked.

I'd stick with GPT when possible, but that's just personal preference. The image is not booting, but I'm not sure if that is related with GPT or MBR, as U-Boot is loaded without issue:
```
DDR Version V1.10 20210810
In
ddrconfig:7
LPDDR4X, 324MHz
BW=32 Col=10 Bk=8 CS0 Row=16 CS1 Row=16 CS=2 Die BW=16 Size=4096MB
tdqss: cs0 dqs0: -24ps, dqs1: -96ps, dqs2: -72ps, dqs3: -144ps, 
tdqss: cs1 dqs0: 48ps, dqs1: -72ps, dqs2: -48ps, dqs3: -120ps, 

change to: 324MHz
PHY drv:clk:36,ca:36,DQ:29,odt:60
vrefinner:16%, vrefout:41%
dram drv:40,odt:0
clk skew:0x61

change to: 528MHz
PHY drv:clk:36,ca:36,DQ:29,odt:60
vrefinner:16%, vrefout:41%
dram drv:40,odt:0
clk skew:0x58

change to: 780MHz
PHY drv:clk:36,ca:36,DQ:29,odt:60
vrefinner:16%, vrefout:41%
dram drv:40,odt:0
clk skew:0x58

change to: 1560MHz(final freq)
PHY drv:clk:36,ca:36,DQ:29,odt:60
vrefinner:16%, vrefout:22%
dram drv:40,odt:80
vref_ca:00000071
clk skew:0x1c
cs 0:
the read training result:
DQS0:0x36, DQS1:0x34, DQS2:0x35, DQS3:0x33, 
min  :0x11 0x12 0x14  0xf  0x1  0x8  0x9  0x7 , 0xb  0xa  0x4  0x1  0xd  0xc  0xd  0x8 ,
      0x12 0x12  0xb  0xb  0x4  0x2  0x2  0x5 , 0xf  0xa  0x9  0x1 0x12  0xf  0xf 0x12 ,
mid  :0x2d 0x2e 0x30 0x2b 0x1e 0x23 0x25 0x22 ,0x26 0x26 0x1e 0x1c 0x29 0x27 0x27 0x24 ,
      0x2e 0x2e 0x27 0x26 0x1f 0x1d 0x1e 0x20 ,0x2a 0x25 0x25 0x1d 0x2c 0x2b 0x2a 0x2c ,
max  :0x49 0x4a 0x4c 0x47 0x3b 0x3f 0x41 0x3d ,0x42 0x42 0x38 0x38 0x45 0x42 0x42 0x40 ,
      0x4b 0x4b 0x44 0x42 0x3b 0x38 0x3a 0x3b ,0x45 0x41 0x42 0x3a 0x47 0x47 0x45 0x47 ,
range:0x38 0x38 0x38 0x38 0x3a 0x37 0x38 0x36 ,0x37 0x38 0x34 0x37 0x38 0x36 0x35 0x38 ,
      0x39 0x39 0x39 0x37 0x37 0x36 0x38 0x36 ,0x36 0x37 0x39 0x39 0x35 0x38 0x36 0x35 ,
the write training result:
DQS0:0x18, DQS1:0x9, DQS2:0xe, DQS3:0x0, 
min  :0x68 0x6a 0x6d 0x69 0x58 0x5e 0x5f 0x60 0x62 ,0x54 0x52 0x4d 0x4f 0x55 0x56 0x57 0x56 0x50 ,
      0x61 0x5c 0x5a 0x57 0x4e 0x4e 0x4e 0x52 0x54 ,0x4c 0x4d 0x49 0x48 0x51 0x50 0x4d 0x54 0x48 ,
mid  :0x82 0x86 0x88 0x83 0x73 0x78 0x7a 0x78 0x79 ,0x6d 0x6d 0x65 0x67 0x70 0x6e 0x6f 0x6e 0x6a ,
      0x7c 0x79 0x74 0x72 0x6a 0x68 0x69 0x6c 0x6f ,0x67 0x67 0x62 0x5f 0x6b 0x6b 0x67 0x6d 0x62 ,
max  :0x9d 0xa2 0xa4 0x9d 0x8f 0x93 0x95 0x91 0x90 ,0x87 0x89 0x7d 0x7f 0x8c 0x87 0x87 0x87 0x84 ,
      0x97 0x96 0x8e 0x8e 0x86 0x82 0x84 0x87 0x8a ,0x83 0x82 0x7c 0x76 0x86 0x87 0x81 0x87 0x7d ,
range:0x35 0x38 0x37 0x34 0x37 0x35 0x36 0x31 0x2e ,0x33 0x37 0x30 0x30 0x37 0x31 0x30 0x31 0x34 ,
      0x36 0x3a 0x34 0x37 0x38 0x34 0x36 0x35 0x36 ,0x37 0x35 0x33 0x2e 0x35 0x37 0x34 0x33 0x35 ,
cs 1:
the read training result:
DQS0:0x35, DQS1:0x35, DQS2:0x34, DQS3:0x2f, 
min  :0x12 0x13 0x14  0xf  0x2  0x8  0xa  0x7 , 0xa  0xa  0x4  0x3  0xe  0xb  0xe  0x9 ,
      0x12 0x12  0xd  0xb  0x4  0x2  0x3  0x7 , 0xf  0x9  0x9  0x3 0x10  0xf  0xd 0x11 ,
mid  :0x2d 0x2e 0x2f 0x2b 0x1e 0x24 0x25 0x21 ,0x27 0x26 0x1f 0x1f 0x2b 0x28 0x29 0x25 ,
      0x2d 0x2d 0x28 0x25 0x1f 0x1d 0x1e 0x20 ,0x28 0x23 0x23 0x1c 0x2a 0x29 0x27 0x2a ,
max  :0x49 0x49 0x4b 0x47 0x3a 0x40 0x41 0x3c ,0x44 0x43 0x3b 0x3b 0x48 0x45 0x45 0x42 ,
      0x48 0x48 0x44 0x40 0x3b 0x38 0x39 0x3a ,0x41 0x3d 0x3d 0x36 0x44 0x44 0x42 0x44 ,
range:0x37 0x36 0x37 0x38 0x38 0x38 0x37 0x35 ,0x3a 0x39 0x37 0x38 0x3a 0x3a 0x37 0x39 ,
      0x36 0x36 0x37 0x35 0x37 0x36 0x36 0x33 ,0x32 0x34 0x34 0x33 0x34 0x35 0x35 0x33 ,
the write training result:
DQS0:0x18, DQS1:0x9, DQS2:0xe, DQS3:0x0, 
min  :0x76 0x77 0x79 0x77 0x67 0x6d 0x6f 0x6e 0x6f ,0x59 0x57 0x50 0x52 0x59 0x59 0x59 0x58 0x53 ,
      0x67 0x66 0x61 0x5f 0x56 0x56 0x56 0x59 0x5a ,0x53 0x52 0x52 0x4d 0x57 0x59 0x53 0x59 0x50 ,
mid  :0x91 0x93 0x95 0x91 0x82 0x87 0x89 0x87 0x87 ,0x72 0x72 0x6a 0x6a 0x74 0x72 0x73 0x72 0x6f ,
      0x82 0x81 0x7b 0x7a 0x71 0x6f 0x70 0x73 0x75 ,0x6f 0x6d 0x6c 0x67 0x72 0x74 0x6e 0x74 0x69 ,
max  :0xad 0xaf 0xb1 0xac 0x9d 0xa2 0xa4 0xa0 0x9f ,0x8c 0x8e 0x84 0x83 0x8f 0x8c 0x8e 0x8c 0x8b ,
      0x9e 0x9c 0x95 0x96 0x8d 0x89 0x8a 0x8e 0x91 ,0x8b 0x89 0x86 0x81 0x8d 0x8f 0x8a 0x8f 0x82 ,
range:0x37 0x38 0x38 0x35 0x36 0x35 0x35 0x32 0x30 ,0x33 0x37 0x34 0x31 0x36 0x33 0x35 0x34 0x38 ,
      0x37 0x36 0x34 0x37 0x37 0x33 0x34 0x35 0x37 ,0x38 0x37 0x34 0x34 0x36 0x36 0x37 0x36 0x32 ,
CA Training result:
cs:0 min  :0x4f 0x49 0x45 0x39 0x45 0x35 0x4c ,0x4e 0x45 0x45 0x37 0x41 0x37 0x4d ,
cs:0 mid  :0x89 0x8a 0x7d 0x7c 0x7e 0x78 0x78 ,0x89 0x86 0x7d 0x7b 0x7b 0x78 0x7c ,
cs:0 max  :0xc4 0xcc 0xb6 0xbf 0xb7 0xbb 0xa4 ,0xc4 0xc7 0xb6 0xbf 0xb5 0xba 0xab ,
cs:0 range:0x75 0x83 0x71 0x86 0x72 0x86 0x58 ,0x76 0x82 0x71 0x88 0x74 0x83 0x5e ,
cs:1 min  :0x4c 0x4a 0x43 0x3e 0x43 0x3b 0x4b ,0x4d 0x48 0x41 0x3e 0x40 0x3a 0x4d ,
cs:1 mid  :0x89 0x8a 0x7f 0x7e 0x7f 0x79 0x7b ,0x8a 0x85 0x7b 0x7a 0x7a 0x78 0x7a ,
cs:1 max  :0xc7 0xcb 0xbb 0xbe 0xbb 0xb7 0xab ,0xc7 0xc2 0xb6 0xb7 0xb5 0xb7 0xa8 ,
cs:1 range:0x7b 0x81 0x78 0x80 0x78 0x7c 0x60 ,0x7a 0x7a 0x75 0x79 0x75 0x7d 0x5b ,
out
U-Boot SPL board init
U-Boot SPL 2017.09-armbian (Feb 17 2023 - 22:34:20)
unknown raw ID phN
unrecognized JEDEC id bytes: 00, 00, 00
Trying to boot from MMC2
## Verified-boot: 0
## Checking atf-1 0x00040000 ... sha256(2f01bd8955...) + OK
## Checking uboot 0x00a00000 ... sha256(ab53d4f780...) + OK
## Checking fdt 0x00b2ecd8 ... sha256(4bdde2de3f...) + OK
## Checking atf-2 0xfdcc9000 ... sha256(f1fecab971...) + OK
## Checking atf-3 0xfdcd0000 ... sha256(d7aa45eb18...) + OK
Jumping to U-Boot(0x00a00000) via ARM Trusted Firmware(0x00040000)
Total: 236.494 ms

INFO:    Preloader serial: 2
NOTICE:  BL31: v2.3():v2.3-152-g4e725b15f:cl
NOTICE:  BL31: Built : 10:51:13, Jul 15 2021
INFO:    GICv3 without legacy support detected.
INFO:    ARM GICv3 driver initialized in EL3
INFO:    pmu v1 is valid
INFO:    dfs DDR fsp_param[0].freq_mhz= 1560MHz
INFO:    dfs DDR fsp_param[1].freq_mhz= 324MHz
INFO:    dfs DDR fsp_param[2].freq_mhz= 528MHz
INFO:    dfs DDR fsp_param[3].freq_mhz= 780MHz
INFO:    Using opteed sec cpu_context!
INFO:    boot cpu mask: 0
INFO:    BL31: Initializing runtime services
WARNING: No OPTEE provided by BL2 boot loader, Booting device without OPTEE initialization. SMC`s destined for OPTEE will return SMC_UNK
ERROR:   Error initializing runtime service opteed_fast
INFO:    BL31: Preparing for EL3 exit to normal world
INFO:    Entry point address = 0xa00000
INFO:    SPSR = 0x3c9


U-Boot 2017.09-armbian (Feb 17 2023 - 22:34:20 +0000)

Model: Radxa ROCK3 Model A
PreSerial: 2, raw, 0xfe660000
DRAM:  4 GiB
Sysmem: init
Relocation Offset: ed349000
Relocation fdt: eb9f69a0 - eb9fecc8
CR: M/C/I
Using default environment

PCIe-0 Link Fail
no mmc device at slot 1
dwmmc@fe2b0000: 1 (SD), dwmmc@fe2c0000: 2, sdhci@fe310000: 0
Bootdev(atags): mmc 1
MMC1: Legacy, 52Mhz
PartType: EFI
No misc partition
boot mode: None
FIT: No boot partition
No resource partition
No resource partition
Failed to load DTB, ret=-19
Failed to get kernel dtb, ret=-19
I2c0 speed: 100000Hz
vsel-gpios- not found! Error: -2
vdd_cpu init 900000 uV
PMIC:  RK8090 (on=0x40, off=0x00)
vdd_logic init 900000 uV
vdd_gpu init 900000 uV
vdd_npu init 900000 uV
io-domain: OK
Model: Radxa ROCK3 Model A
rockchip_set_serialno: could not find efuse/otp device
CLK: (sync kernel. arm: enter 816000 KHz, init 816000 KHz, kernel 0N/A)
  apll 816000 KHz
  dpll 780000 KHz
  gpll 1188000 KHz
  cpll 1000000 KHz
  npll 1200000 KHz
  vpll 24000 KHz
  hpll 24000 KHz
  ppll 200000 KHz
  armclk 816000 KHz
  aclk_bus 150000 KHz
  pclk_bus 100000 KHz
  aclk_top_high 300000 KHz
  aclk_top_low 200000 KHz
  hclk_top 150000 KHz
  pclk_top 100000 KHz
  aclk_perimid 300000 KHz
  hclk_perimid 150000 KHz
  pclk_pmu 100000 KHz
No misc partition
Net:   eth1: ethernet@fe010000
Hit key to stop autoboot('CTRL+C'):  0 
failed to find reset-gpios property

Device 0: unknown device
switch to partitions #0, OK
mmc1 is current device
Failed to mount ext2 filesystem...
** Unrecognized filesystem type **
MMC error: The cmd index is 0, ret is -110
mmc_init: -110, time 4
starting USB...
Bus dwc3@fcc00000: usb maximum-speed not found
Register 2000140 NbrPorts 2
Starting the controller
USB XHCI 1.10
Bus dwc3@fd000000: usb maximum-speed not found
Register 2000140 NbrPorts 2
Starting the controller
USB XHCI 1.10
scanning bus dwc3@fcc00000 for devices... 1 USB Device(s) found
scanning bus dwc3@fd000000 for devices... 1 USB Device(s) found
       scanning usb for storage devices... 0 Storage Device(s) found

Device 0: unknown device
failed to find reset-gpios property
ethernet@fe010000 Waiting for PHY auto negotiation to complete.
```

**Reply by MichaIng** [reply]
> I'd stick with GPT when possible, but that's just personal preference.

It is the future for sure, but MBR is still easier to handle, no secondary partition table to deal with (one less needed tool), less data overhead etc. So until > 2 TiB becomes a thing for SD cards, eMMC or flash drives, I'll stick with MBR where possible.

Thanks for the serial console output. That looks exactly what I saw on Orange Pi 5, so GPT only. I'll create a new image with single ext4 partition but GPT partition table.

**Reply by MichaIng** [reply]
New GPT + single ext4 image is here btw, would be great if someone could do a final test, so I can move it to stable downloads: https://dietpi.com/downloads/images/testing/

**Reply by RadxaYuntian** [reply]
The last modified date for 3A image is still 2023-03-24 11:02. Are you sure it was updated?

**Reply by MichaIng** [reply]
Yes, I updated the image right after your negative result with the MBR partition table.

**Reply by RadxaYuntian** [reply]
I was able to boot the kernel, but there was only kernel message over the serial, no login console. HDMI also had no output.

**Reply by MichaIng** [reply]
Hmm, this cannot have something to do with the partitioning if the kernel is booting already. Is it the same with the current stable image from [here](https://dietpi.com/downloads/images/)?

I compared both, and everything (relevant) seems to be identical apart of partitioning, especially kernel version and boot configs are the same. Also a login console should be enabled on ttyS2 (being the default debug console port, right?). Note that it may take a while before the login prompt appears on serial console due to first boot setup steps which are shown only on the main console (tty1), i.e. on HDMI. To have all boot messages on serial console, one needs to switch the two `console=` values in `/boot/dietpiEnv.txt` `consoleargs=` line. The one named last is the one one used for all init system logs.

**Reply by MichaIng** [reply]
New images built: https://dietpi.com/downloads/images/testing/

**Reply by wiedehopf** [reply]
Hey do you still need testing on this?
I've got one on the way as they were on sale for 20 euros with shipping: `https://www.amazon.de/dp/B09NFQG69F`

**Reply by MichaIng** [reply]
It should work just fine. I have not heard of any issues and we have quite a bunch of recent ROCK 3A systems running DietPi reported.

**Comment by Deoptim** [comment]
Hello.
Is there any possibility to request Vendor kernel 6.1.*?
Only NPU, MPP+RGA(ffmpeg) and Frigate works under Vendor kernel if installed to Home Assistant.
Thanks.

**Reply by MichaIng** [reply]
Can you try this one:
```sh
cd /tmp
wget https://dietpi.com/downloads/binaries/linux-{dtb,image}-vendor-rk35xx.deb
dpkg -i linux-{dtb,image}-vendor-rk35xx.deb
reboot
```
I hope mainline U-Boot does not somehow have issues with the vendor kernel. At best apply only when you can access the SD card on another system to point the Image/dtb/uInitrd symlinks back to mainline kernel, if needed.
